text/0314-cdk-construct-hub.md

```diff
@@ -362,6 +367,96 @@ pipeline.
 The details of URLs to be configrued in CloudFront is available in [appendix
 "CloudFront URLs"](#cloudfront-urls).
 
+#### Backend X Frontend Compatibility
+
+As described earlier, the front-end application consumes artifacts that are stored on the backend. Since these artifacts are subject to change,
+we need to define a mechanism by which the front-end is able to detect and operate on various versions of them.
+
+To that end, we propose a generic mechanism to apply to all artifact types:
+
+Every artifact will comply to the [SemVer](https://semver.org/spec/v2.0.0.html) specification
+and include its major version in the `version` field of the schema:
+
+```json
+{
+  "version": "v1",
+  ...
+}
+```
+
+```json
+{
+  "version": "v2",
+  ...
+}
+```
+
+Every time a change in the artifact occurs, the contents of the artifact is replaced with the new version.
+This means that at any given point in time, the artifact can either be in its new version, or the old one.
+
+As a general guideline, we *prefer displaying an older version of an artifact, than to display nothing at all.*.
+To that end, the front-end application will have to support displaying all versions of the artifact.
```

Comment:
It'll need to be able to render documents using a set of previous versions of the schema, but it does not have to be _all_ versions. Generally speaking, we could require that instances are upgraded through all interim major versions... So the front-end of any major may only ever need to render the current & previous version.

I guess this might be the spirit of what you were trying to say on L395, but the fact the statement does not specifically call out *major* versions, it's a little bit blurry.

Reply:
> Generally speaking, we could require that instances are upgraded through all interim major versions... So the front-end of any major may only ever need to render the current & previous version.

How would we do that? Imagine a breaking change is deployed, meaning the front-end now has to support two major versions. Now, before the backfilling finishes, another breaking change is deployed, so some packages may have major version 3, but some might still be at major version 2. The front-end now has to support 3 major versions.

Reply:
My argument is that you could _require_ all objects be migrated to v2 before the v3 change can be deployed. There could be a deploy-time pre-condition check, for example.

Reply:
I see, but do we want that? I think if we see it becomes a burden for the front-end to support, maybe introduce this option in the future.